### PR TITLE
Pushover Sink

### PR DESCRIPTION
**Pull request by Nabsku**
Because I wanted this myself and I couldn't get it to work with the webhook, here is a really really rudimentary version of a pushover sink. It works with log and image enrichments but supports only one image attachment per message and a maximum message size of 1024 characters.

Please excuse the poor programming, just getting started :)

Opening this as a draft, so that I can get more eyes on this (hope thats okay?).

### CODE REVIEW

File: src/robusta/core/sinks/pushover/pushover_sink.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                                     sink_config.pushover_sink.device)
+        
+        self.send_files = sink_config.pushover_sink.send_files
+        self.send_as_html = sink_config.pushover_sink.send_as_html
+
+    def write_finding(self, finding: Finding, platform_enabled: bool):
+        self.__send_pushover_message(finding, platform_enabled)
+
+    def __send_pushover_message(self, finding: Finding, platform_enabled: bool):
+        if self.send_files:
```
Comment (arikalon1) [inline src/robusta/core/sinks/pushover/pushover_sink.py:38]: I think this function could be simplified to the below. Basically, first send the finding details (type of issue, related resource, severity etc), the the other blocks
```
    def __send_pushover_message(self, finding: Finding, platform_enabled: bool):
        # first send finding data
        title, message = self.__get_message_text(finding, platform_enabled)
        investigate_url = finding.get_investigate_uri(self.account_id, self.cluster_name) if platform_enabled else ""
        self.client.send_message(
            title=title, message=message, send_as_html=self.send_as_html, additional_url=investigate_url
        )

        if self.send_files:
            for enrichment in finding.enrichments:
                file_blocks = [block for block in enrichment.blocks if isinstance(block, FileBlock)]
                image_blocks = [block for block in file_blocks if is_image(block.filename)]
                text_blocks = [block for block in file_blocks if block.filename.endswith((".txt", ".log"))]

                if len(image_blocks) > 0:
                    for block in image_blocks:
                        self.client.send_file(
                            file_name=block.filename,
                            contents=block.contents,
                            title=title,
                            message=message,
                            send_as_html=self.send_as_html,
                            additional_url=investigate_url,
                        )

                if len(text_blocks) > 0:
                    for block in text_blocks:
                        message += f"<b>Additional content:</b> \n\n {block.contents.decode()}"

                        self.client.send_message(
                            title=title,
                            message=(message[:1024] if len(message) > 1024 else message),
                            send_as_html=self.send_as_html,
                            additional_url=investigate_url,
                        )

```

**Review by arikalon1 — COMMENTED** [review]
Thanks for the PR @Nabsku
Nice work!
Left few comments.

Since there's a size limitation of only 1024 bytes, I assume many notifications will be trimmed.
Does this limit applies to files as well?
Does pushover also accept text file attachments? (because you sent text files as text and not attachment)

I think I would consider sending only the finding details, and not enrichments. 
(type of issue, severity, k8s resources, description)
Leave a link to explore all the data on the platform.

Thoughts?

File: src/robusta/core/sinks/pushover/pushover_client.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import logging
+import os
+
+import requests
+
+from robusta.core.reporting.utils import PNG_SUFFIX, SVG_SUFFIX, convert_svg_to_png, is_image
+
+PUSHOVER_BASE_URL = os.environ.get("PUSHOVER_BASE_URL", "https://api.pushover.net/1/messages.json")
```
[the remark this answers is not in the document]
Comment (Nabsku) [reply src/robusta/core/sinks/pushover/pushover_client.py:8]: Good catch, was just following the telegram sink on this one

File: src/robusta/core/sinks/pushover/pushover_client.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            "message": message,
+            "url": additional_url,
+            "device": self.device
+        }
+
+        response = requests.post(url, json=message_json)
+
+        if response.status_code != 200:
+            logging.error(
+                f"Failed to send pushover message: token - {self.token} reason - {response.reason} {response.text}"
```
[the remark this answers is not in the document]
Comment (Nabsku) [reply src/robusta/core/sinks/pushover/pushover_client.py:33]: Sorry, yeah - of course. Those were debugging statements, will be removed!

**Comment by Nabsku** [comment]
> Since there's a size limitation of only 1024 bytes, I assume many notifications will be trimmed. Does this limit applies to files as well? Does pushover also accept text file attachments? (because you sent text files as text and not attachment)

Pushover does only accept text as normal message input and images as attachments AFAICT. Most longer messages will be trimmed to 1024 chars but I feel like it's better than nothing. Could also think about sending messages in parts, eg part1 with the first 1024 chars and part2 with second batch of 1024 chars and so on. Thoughts on that?

> I think I would consider sending only the finding details, and not enrichments. (type of issue, severity, k8s resources, description) Leave a link to explore all the data on the platform.

Yeah I was considering only sending the details but figured it was a try to get **_some_** enrichments over to Pushover, since I quite like those and they might help you spot the error in an instant, even if you can't display the whole enrichment message. 

What about a toggle? Let the user decide?

And thanks for the feedback!

File: src/robusta/core/sinks/pushover/pushover_sink.py (hunk 1 above)
Comment (Nabsku) [reply src/robusta/core/sinks/pushover/pushover_sink.py:38]: That looks much nicer and cleaner. Let me try this.

**Review by arikalon1 — APPROVED** [review]
Looks good
Thanks for the PR @Nabsku Nabsku

**Comment by arikalon1** [comment]
I think writing the notification in chunks would be a good user experience. 
Having a toggle that enables/disables sending enrichments would be good I think.

I'm merging the PR.
If you also think a toggle would be good, please open a new PR for it.

Thanks for the contribution @Nabsku !

**Comment by Nabsku** [comment]
> I think writing the notification in chunks would be a good user experience. Having a toggle that enables/disables sending enrichments would be good I think.
> 
> I'm merging the PR. If you also think a toggle would be good, please open a new PR for it.
> 
> Thanks for the contribution @Nabsku !

I will look into this shortly. Thanks for guiding me through this process and your work on Robusta!

**Pull request merged**